File: mobile/swift/mobile-swift-swiftui/Sources/Design/TestId.swift
import SwiftUI

extension View {
    /// Leaf testid — text, field, button. The same string the web cell puts in
    /// `data-testid` and the Compose cell puts in `contentDescription`, so
    /// XCUITest / Appium `accessibility id` drives all three cells.
    func testId(_ id: String) -> some View {
        accessibilityIdentifier(id)
    }

    /// Wrapper testid (`header`, `login-form`, `welcome-panel`, …). Keeps the
    /// container addressable while its children stay individually queryable —
    /// mirrors `data-testid` on a wrapper element.
    func containerTestId(_ id: String) -> some View {
        accessibilityElement(children: .contain)
            .accessibilityIdentifier(id)
    }

    /// Same as ``testId(_:)`` for slots whose id is optional (panel titles).
    @ViewBuilder
    func optionalTestId(_ id: String?) -> some View {
        if let id {
            accessibilityIdentifier(id)
        } else {
            self
        }
    }
}

File: frontend/kotlin/frontend-kotlin-compose/app/src/main/kotlin/dev/multistack/compose/ui/TestId.kt
package dev.multistack.compose.ui

import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics

/**
 * One testid string, two Appium locators — the same value the web cell puts in
 * `data-testid`:
 *  - `content-desc` (Appium `accessibility id`) — cross-platform selector shared
 *    with the SwiftUI cell's `accessibilityIdentifier`
 *  - `resource-id` — via `testTag`, exported because the root sets
 *    `testTagsAsResourceId = true` (see [MultistackRoot])
 *
 * Semantics are **not** merged, so a container keeps its own id while children
 * stay individually findable (mirrors `data-testid` on a wrapper element).
 */
fun Modifier.testId(id: String): Modifier = this
    .testTag(id)
    .semantics { contentDescription = id }
